src/pw/fpga/opencl_utils.c

```diff
@@ -0,0 +1,370 @@
+
+  void *memptr = NULL;
+  int ret = posix_memalign(&memptr, OPENCL_ALIGNMENT, size);
+  return memptr;
+}
+
+void openCLContextCallBackFxn(const char *errinfo, const void *private_info, size_t cb, void *user_data) {
+  printf("Context Callback - %s\n", errinfo);
+}
+
+void printError(cl_int error) {
+
+  switch(error)
+  {
+    case CL_INVALID_PLATFORM:
+      printf("CL_PLATFORM NOT FOUND OR INVALID ");
+      break;
+    case CL_INVALID_DEVICE:
+      printf("CL_DEVICE NOT FOUND OR INVALID OR DOESN'T MATCH THE PLATFORM ");
+      break;
+    case CL_INVALID_CONTEXT:
+      printf("CL_CONTEXT INVALID ");
+      break;
+    case CL_OUT_OF_HOST_MEMORY:
+      printf("FAILURE TO ALLOCATE RESOURCES BY OPENCL");
+      break;
+    case CL_DEVICE_NOT_AVAILABLE:
+      printf("CL_DEVICE NOT AVAILABLE ALTHOUGH FOUND");
+      break;
+    case CL_INVALID_QUEUE_PROPERTIES:
+      printf("CL_QUEUE PROPERTIES INVALID");
+      break;
+    case CL_INVALID_PROGRAM:
+      printf("CL_PROGRAM INVALID");
+      break;
+    case CL_INVALID_BINARY:
+      printf("CL_BINARY INVALID");
+      break;
+    case CL_INVALID_KERNEL_NAME:
+      printf("CL_KERNEL_NAME INVALID");
+      break;
+    case CL_INVALID_KERNEL_DEFINITION:
+      printf("CL_KERNEL_DEFN INVALID");
+      break;
+    case CL_INVALID_VALUE:
+      printf("CL_VALUE INVALID");
+      break;
+    case CL_INVALID_BUFFER_SIZE:
+      printf("CL_BUFFER_SIZE INVALID");
+      break;
+    case CL_INVALID_HOST_PTR:
+      printf("CL_HOST_PTR INVALID");
+      break;
+    case CL_INVALID_COMMAND_QUEUE:
+      printf("CL_COMMAND_QUEUE INVALID");
+      break;
+    case CL_INVALID_MEM_OBJECT:
+      printf("CL_MEM_OBJECT INVALID");
+      break;
+    case CL_MEM_OBJECT_ALLOCATION_FAILURE:
+      printf("CL_MEM_OBJECT_ALLOCATION INVALID");
+      break;
+    case CL_INVALID_ARG_INDEX:
+      printf("CL_ARG_INDEX INVALID");
+      break;
+    case CL_INVALID_ARG_VALUE:
+      printf("CL_ARG_VALUE INVALID");
+      break;
+    case CL_INVALID_ARG_SIZE:
+      printf("CL_ARG_SIZE INVALID");
+      break;
+    case CL_INVALID_PROGRAM_EXECUTABLE:
+      printf("CL_PROGRAM_EXEC INVALID");
+      break;
+    case CL_INVALID_KERNEL:
+      printf("CL_KERNEL INVALID");
+      break;
+    case CL_INVALID_KERNEL_ARGS:
+      printf("CL_KERNEL_ARG INVALID");
+      break;
+    case CL_INVALID_WORK_GROUP_SIZE:
+      printf("CL_WORK_GROUP_SIZE INVALID");
+      break;
+
+    default:
+      printf("UNKNOWN ERROR %d\n", error);
+  }
+
+}
+
+void _checkError(const char *file, int line, const char *func, cl_int err, const char *msg, ...){
+
+  if(err != CL_SUCCESS){
+    printf("ERROR: ");
+    printError(err);
+    printf("\nError Location: %s:%d:%s\n", file, line, func);
+
+    // custom message 
+    va_list vl;
+    va_start(vl, msg);
+    vprintf(msg, vl);
+    printf("\n");
+    va_end(vl);
+
+    queue_cleanup();
+    cleanup();
+    exit(err);
+  }
+}
+
+static void tolowercase(char *p, char *q){
```

Comment:
please document the assumptions for this function (`p` must be NULL terminated, `q` must hold `strlen(p)+1` characters)

Reply:
Made this change. Thanks